### PR TITLE
refactor(time-input): migrate to typescript

### PR DESCRIPTION
**Pull request by ddouglasz**
#### Summary

- Migrate time input to typescript

### CODE REVIEW

File: packages/components/inputs/time-input/src/time-input-body.tsx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import PropTypes from 'prop-types';
+import { KeyboardEvent, MouseEvent } from 'react';
```
Comment (Rhotimee) [inline packages/components/inputs/time-input/src/time-input-body.tsx:1]:
```suggestion
import type { KeyboardEvent, MouseEvent } from 'react';
```

File: packages/components/inputs/time-input/src/time-input-body.tsx [hunk 2]
```diff
@@ -13,35 +13,46 @@ import {
   StyledInputContainer,
 } from './time-input-body.styles';
 
-export const ClearSection = (props) => {
+import { TTimeInput } from './time-input';
```
Comment (Rhotimee) [inline packages/components/inputs/time-input/src/time-input-body.tsx:16]:
```suggestion
import type { TTimeInput } from './time-input';
```

File: packages/components/inputs/time-input/src/time-input.tsx [hunk 3]
```diff
@@ -11,21 +10,118 @@ import { useFieldId, usePrevious } from '@commercetools-uikit/hooks';
...
+   */
+  name?: string;
+  /**
+   * Used as HTML autocomplete of the input component.
+   */
+  autoComplete?: string;
+  /**
+   * Value of the input
+   */
+  value: string;
```
Comment (Rhotimee) [inline packages/components/inputs/time-input/src/time-input.tsx:56]:
```suggestion
  value?: string;
```

File: .changeset/twenty-lions-glow.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@commercetools-uikit/time-input': major
```
Comment (Rhotimee) [inline .changeset/twenty-lions-glow.md:2]: I think this should be a patch

**Review by Rhotimee — CHANGES_REQUESTED** [review]

File: packages/components/inputs/time-input/src/time-input-body.tsx (hunk 1 above)
Comment (tdeekens) [reply packages/components/inputs/time-input/src/time-input-body.tsx:1]: 👋🏼 Just for completeness depending on the TypeScript version one can now use `import { type foo } from './bar'` which means one doesn't have to split type imports.

I think we can stick with `import type` here and eventually on a Friday move everything to `import { type foo }` on a Friday. Did the same on `flopflip` [here](https://github.com/tdeekens/flopflip/pull/1558/files).

🔗 https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/#type-on-import-names

File: packages/components/inputs/time-input/src/time-input.tsx [hunk 5]
```diff
@@ -11,21 +10,118 @@ import { useFieldId, usePrevious } from '@commercetools-uikit/hooks';
...
+    | 9
+    | 10
+    | 11
+    | 12
+    | 13
+    | 14
+    | 15
+    | 16
+    | 'scale'
+    | 'auto';
```
Comment (Rhotimee) [inline packages/components/inputs/time-input/src/time-input.tsx:44]: The initial horizontalConstraint prop option starts from 1. 

```suggestion
  horizontalConstraint?:
    | 1
    | 2
    | 3
    | 4
    | 5
    | 6
    | 7
    | 8
    | 9
    | 10
    | 11
    | 12
    | 13
    | 14
    | 15
    | 16
    | 'scale'
    | 'auto';
```

**Review by Rhotimee — APPROVED** [review]

File: packages/components/inputs/time-input/src/time-input.tsx [hunk 6]
```diff
@@ -11,21 +10,120 @@ import { useFieldId, usePrevious } from '@commercetools-uikit/hooks';
...
+   * <br/>
+   * Signature: `(event) => void`
+   */
+  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
+  /**
+   * Called when input is focused
+   * <br/>
+   * Signature: `(event) => void`
+   */
+  onFocus?: () => void;
```
Comment (Rhotimee) [inline packages/components/inputs/time-input/src/time-input.tsx:78]: I think we can improve the event handlers a bit. 

```suggestion
type TEvent = {
  target: {id: string; name?: string; value: unknown}
}
  onChange: (event: TEvent) => void;
  /**
   * Called when input is blurred
   * <br/>
   * Signature: `(event) => void`
   */
  onBlur?: FocusEventHandler
  /**
   * Called when input is focused
   * <br/>
   * Signature: `(event) => void`
   */
  onFocus?: FocusEventHandler
```

File: packages/components/inputs/time-input/src/time-input-body.tsx [hunk 7]
```diff
@@ -1,4 +1,4 @@
-import PropTypes from 'prop-types';
+import { type KeyboardEvent, type MouseEvent } from 'react';
```
Comment (Rhotimee) [inline packages/components/inputs/time-input/src/time-input-body.tsx:1]:
```suggestion
import type { KeyboardEvent, MouseEvent } from 'react';
```

File: packages/components/inputs/time-input/src/time-input.tsx (hunk 5 above)
Comment (Rhotimee) [reply packages/components/inputs/time-input/src/time-input.tsx]: @ddouglasz, I made a mistake. This should actually start from 3. As that was the initial values.
It's time-body-input that starts from 1. 
Sorry about that.

**Review by Rhotimee — CHANGES_REQUESTED** [review]
@ddouglasz there are some unrelated changes here. Please check it out.

**Comment by ddouglasz** [comment]
> @ddouglasz can you do a rebase here? It's a bit hard to review only the relevant changes. Thanks

I am currently working on it. Thank you.

File: packages/components/inputs/time-input/src/time-input.tsx [hunk 8]
```diff
@@ -11,21 +10,122 @@ import { useFieldId, usePrevious } from '@commercetools-uikit/hooks';
...
+  minutes: number;
+  seconds: number;
+  milliseconds: number;
+};
+
+type TEvent = {
+  target: { id: string; name?: string; value: unknown };
+};
+
+export type TTimeInput = {
```
Comment (emmenko) [inline packages/components/inputs/time-input/src/time-input.tsx:24]:
```suggestion
export type TTimeInputProps = {
```

File: packages/components/inputs/time-input/src/time-input.tsx [hunk 9]
```diff
@@ -11,21 +10,122 @@ import { useFieldId, usePrevious } from '@commercetools-uikit/hooks';
...
+  /**
+   * Value of the input
+   */
+  value?: string;
+  /**
+   * Called with an event holding the new value.
+   * <br/>
+   * Required when input is not read only. Parent should pass it back as `value`-
+   * <br />
+   * Signature: `(event) => void`
```
Comment (emmenko) [inline packages/components/inputs/time-input/src/time-input.tsx:66]: These `Signature` comments can be removed

File: packages/components/inputs/time-input/src/time-input.tsx [hunk 10]
```diff
@@ -212,7 +235,7 @@ TimeInput.toLocaleTime = (time, locale) => {
     minute: 'numeric',
     // only show seconds when time contains seconds
     second: parsedTime.seconds > 0 ? 'numeric' : undefined,
-  };
+  } as const;
```
Comment (emmenko) [inline packages/components/inputs/time-input/src/time-input.tsx:238]: What's this for?

File: packages/components/inputs/time-input/src/time-input.tsx [hunk 11]
```diff
@@ -45,7 +145,10 @@ const TimeInput = (props) => {
   const handleBlur = useCallback(
     (event) => {
       // check formatting and reformat when necessary
-      const formattedTime = TimeInput.toLocaleTime(value, intl.locale);
+      const formattedTime = TimeInput.toLocaleTime(
+        value as string,
```
Comment (emmenko) [inline packages/components/inputs/time-input/src/time-input.tsx:149]: Is casting necessary here?

File: packages/components/inputs/time-input/src/time-input-body.tsx [hunk 12]
```diff
@@ -13,35 +13,46 @@ import {
   StyledInputContainer,
 } from './time-input-body.styles';
 
-export const ClearSection = (props) => {
+import type { TTimeInput } from './time-input';
+
+type TExtendedTimeInput = TTimeInput & {
```
Comment (emmenko) [inline packages/components/inputs/time-input/src/time-input-body.tsx:18]:
```suggestion
type TTimeInputBodyProps = TTimeInput & {
```

File: packages/components/inputs/time-input/src/time-input-body.tsx [hunk 13]
```diff
@@ -13,35 +13,46 @@ import {
...
+  onClear: (
+    event: MouseEvent<HTMLButtonElement> | KeyboardEvent<HTMLButtonElement>
+  ) => void;
+};
+
+type TClearSection = {
```
Comment (emmenko) [inline packages/components/inputs/time-input/src/time-input-body.tsx:24]:
```suggestion
type TClearSectionProps = {
```

File: packages/components/inputs/time-input/src/time-input.tsx (hunk 10 above)
Comment (ddouglasz) [reply packages/components/inputs/time-input/src/time-input.tsx:238]: I was looking at narrowing down the literal types from the `options` expression using const assertion. I think I can use the types from `Intl.DateTimeFormatOptions` as the `options ` type definition instead.

**Review by emmenko — APPROVED** [review]

**Review by Rhotimee — APPROVED** [review]

**Pull request merged**